Server WMS GetPrint accesscontrol support

elpaso opened the pull request:
Fixes #41800

Funded by Gis3W https://www.gis3w.it

elpaso committed:
Fix server getPrint accesscontrol for map items

Fixes #41800

elpaso committed:
Server access control GetPrint support

Fixes #41800

elpaso committed:
Fix deprecated toSet

nyalldawson commented on src/core/layout/qgslayoutitemattributetable.cpp:
```diff
@@ -693,6 +705,48 @@ QVariant QgsLayoutItemAttributeTable::replaceWrapChar( const QVariant &variant )
+  for ( const auto &c : qgis::as_const( allowedColumns ) )
+  {
+    if ( ! c.attribute().isEmpty() )
+      allowedAttributes.insert( c.attribute() );
```
What about expression based columns?

elpaso replied:
The intention was that they pass through untouched. I was convinced that the access control plugins did not handle expression fields but I'll double check it.

nyalldawson replied:
isn't that a potential data leak though? If I had an expression which collated two string fields I'm not supposed to see, you wouldn't want that shown would you?

elpaso replied:
I considered that, If I get this correctly, the `QgsLayoutTableColumn::mAttribute` is not properly named because it actually stores an expression in every case.

Passing the column header makes no sense in this context, so the best we can do here is to pass all `mAttribute`s to the plugin that will decide if the column is visible or not based on its logic.

An additional safety measure would probably be to filter out the forbidden attributes from the feature request directly (this is what Server WFS does).

Note that there are many other potential data leaks in the server (notably: label expression).

elpaso replied:
I've added a test for exp column filters, I didn't need to change the code in any way and all columns `mAttribute`(s) are passed to the plugin including:
- expression columns
- "virtual" fields from the layer

The plugin can decide whether allow them or not so there is a way to hide them.

I think we are good to go with this implementation, I'm not sure if we need to add a lower level filtering at the QgsFeatureRequest level.

elpaso replied:
@nyalldawson I've changed the logic to work on attributes names extracted from the expressions, this should be the right implementation.

I've also fixed atlas support.

nyalldawson commented on src/core/layout/qgslayoutrendercontext.h:
```diff
@@ -297,6 +297,18 @@ class CORE_EXPORT QgsLayoutRenderContext : public QObject
+    QgsFeatureFilterProvider *featureFilterProvider() const;
+
+    /**
+     * Sets \a featureFilterProvider
```
These are very "qgis 2.x" style docs 🙃 Can you add more details here about exactly what this is used for and when it's set?

nyalldawson reviewed (commented):
Can you also lower the dpi of the test reference images to speed up the test comparisons and:
- bump the font size up
- use the qgis test reference font in bold style

Otherwise these tests will be quite fragile

elpaso commented:
> Can you also lower the dpi of the test reference images to speed up the test comparisons

Lower than https://github.com/qgis/QGIS/pull/41823/files#diff-d294e6b04b17a62d3ce6076294f95999279a3adb61825c9fefb0916d3c0ff869R153 ?

Actually I think that the problem is that it is ignored:  https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsrenderer.cpp#L482

I'll fix that,

elpaso committed:
Fix server WMS DPI ignored (unreported)

elpaso committed:
QGIS 3 style doxy

elpaso committed:
Add test for attr table exp filters

elpaso committed:
flake8

elpaso committed:
spell

elpaso committed:
Add tests and fixes for Atlas

The change to the the feature filter group filterFeatures
(do not clear pre-existing filters) should not be a behavioral
change: additional filters are ANDed.

elpaso committed:
Update test images from local docker run

nyalldawson commented on src/core/layout/qgslayoutitemattributetable.cpp:
```diff
@@ -541,6 +551,9 @@ bool QgsLayoutItemAttributeTable::getTableContents( QgsLayoutTableContents &cont
     // corresponding option is enabled)
     QVector< Cell > currentRow;
+#ifdef HAVE_SERVER_PYTHON_PLUGINS
+    mColumns = filteredColumns();
```
If this is permanently replacing the previous column configuration with the filtered columns, won't it mean that columns stripped from one request with limited column visibility which comes before a second request with more permissive configuration are always missing from the second request?

elpaso replied:
@nyalldawson of course the server is supposed to be stateless so we don't have a problem here. I've added a test case to prove it anyway.

nyalldawson commented:
Thanks, reference images look much better now !

Gustry commented:
> Funded by: Gis3W

Not reallly related to the PR and this reminds me I haven't advertise that now. https://changelog.qgis.org will parse "Funded by Gis3W website.com" in the PR description. It will fill automatically the dedicated field and the exported version for the website as RST format.
The only doc for now : https://github.com/kartoza/prj.app/blob/develop/django_project/changes/tests/test_github_pull_request.py#L66

CC @zacharlie 

My 2 cents for other PR ;-)

elpaso committed:
Add unfiltered test case to prove that filters are not persistent

elpaso committed:
typo
